File: app/src/main/java/com/angogasapps/myfamily/utils/BuyListUtils.java
package com.angogasapps.myfamily.utils;

import com.angogasapps.myfamily.models.BuyList;
import com.angogasapps.myfamily.models.Family;
import com.google.firebase.database.DataSnapshot;

import java.util.ArrayList;
import java.util.HashMap;

import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.CHILD_COMMENT;
import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.CHILD_FROM;
import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.CHILD_NAME;
import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.CHILD_PRODUCTS;
import static com.angogasapps.myfamily.firebase.FirebaseVarsAndConsts.USER;

public class BuyListUtils {
    /*
  DataSnapshot {
    key = -MY0xumFvGFkkxbaYiLo,
    value = {
        name=#1,
        products = {
            -MZO2_VpJkWN3pY3fdyE = {
                annotation=,
                count=-1,
                name=продукт 1,
                from=-MZO2_VpJkWN3pY3fdyE,
                id=-MZO2_VpJkWN3pY3fdyE
                }
             }
         }
     }

 */
    public static BuyList parseBuyList(DataSnapshot snapshot){
        BuyList buyList = new BuyList();

        buyList.setId(snapshot.getKey());
        buyList.setName(snapshot.child(CHILD_NAME).getValue(String.class));
        buyList.setProducts(getProductsList(snapshot));

        return buyList;
    }

    private static ArrayList<BuyList.Product> getProductsList(DataSnapshot snapshot){
        ArrayList<BuyList.Product> list = new ArrayList<>();
        for (DataSnapshot productSnapshot : snapshot.child(CHILD_PRODUCTS).getChildren()){
            list.add(getProduct(productSnapshot));
        }

        return list;
    }

    private static BuyList.Product getProduct(DataSnapshot snapshot){
        BuyList.Product product = snapshot.getValue(BuyList.Product.class);
        product.setId(snapshot.getKey());
        return product;
    }

    public static HashMap<String, Object> getHashMap(BuyList.Product product){
        HashMap<String, Object> map = new HashMap<>();
        map.put(CHILD_NAME, product.getName());
        map.put(CHILD_FROM, product.getFrom());
        map.put(CHILD_COMMENT, product.getComment());

        return map;
    }

    public static String getCorrectComment(BuyList.Product product){
        String comment = product.getComment();
        if (product.getFrom().equals(USER.getPhone())){
            if (comment != null){
                return comment;
            }else{
                return "";
            }
        }else {
            if (!comment.equals("")){
                return comment + " : " + Family.getInstance().getNameByPhone(product.getFrom());
            }else{
               return Family.getInstance().getNameByPhone(product.getFrom());
            }
        }
    }

    public static int getIndexOfRemoveProduct(ArrayList<BuyList.Product> oldBuyList, ArrayList<BuyList.Product> newBuyList){
        for (int i = 0; i < newBuyList.size(); i++) {
            if (!oldBuyList.get(i).getId().equals(newBuyList.get(i).getId())){
                return i;
            }
        }
        return newBuyList.size();
    }
}

File: app/src/main/java/com/angogasapps/myfamily/utils/BuyListUtils.kt
package com.angogasapps.myfamily.utils


import com.angogasapps.myfamily.firebase.*
import com.angogasapps.myfamily.models.Family.getNameByPhone
import com.google.firebase.database.DataSnapshot
import com.angogasapps.myfamily.models.buy_list.BuyList
import com.angogasapps.myfamily.utils.BuyListUtils
import com.angogasapps.myfamily.models.buy_list.BuyList.Product
import com.angogasapps.myfamily.models.Family
import java.util.ArrayList
import java.util.HashMap

object BuyListUtils {
    /*
  DataSnapshot {
    key = -MY0xumFvGFkkxbaYiLo,
    value = {
        name=#1,
        products = {
            -MZO2_VpJkWN3pY3fdyE = {
                annotation=,
                count=-1,
                name=продукт 1,
                from=-MZO2_VpJkWN3pY3fdyE,
                id=-MZO2_VpJkWN3pY3fdyE
                }
             }
         }
     }

 */
    fun parseBuyList(snapshot: DataSnapshot): BuyList {
        val buyList = BuyList()
        buyList.id = snapshot.key!!
        buyList.name = snapshot.child(CHILD_NAME).getValue(String::class.java)!!
        buyList.products = getProductsList(snapshot)
        return buyList
    }

    private fun getProductsList(snapshot: DataSnapshot): ArrayList<Product> {
        val list = ArrayList<Product>()
        for (productSnapshot in snapshot.child(CHILD_PRODUCTS).children) {
            list.add(getProduct(productSnapshot)!!)
        }
        return list
    }

    private fun getProduct(snapshot: DataSnapshot): Product? {
        println(snapshot)
        val product = snapshot.getValue(Product::class.java)
        product!!.id = snapshot.key!!
        return product
    }

    fun getHashMap(product: Product): HashMap<String, Any> {
        val map = HashMap<String, Any>()
        map[CHILD_NAME] = product.name
        map[CHILD_FROM] = product.from
        map[CHILD_COMMENT] = product.comment
        return map
    }

    fun getCorrectComment(product: Product): String {
        val comment = product.comment
        return if (product.from == USER.phone) {
            comment ?: ""
        } else {
            if (comment != "") {
                comment + " : " + getNameByPhone(product.from)
            } else {
                getNameByPhone(product.from)
            }
        }
    }

    fun getIndexOfRemoveProduct(
        oldBuyList: ArrayList<Product>,
        newBuyList: ArrayList<Product>
    ): Int {
        for (i in newBuyList.indices) {
            if (oldBuyList[i].id != newBuyList[i].id) {
                return i
            }
        }
        return newBuyList.size
    }

    fun getIndexOfChangeProduct(
        oldBuyList: ArrayList<Product>,
        newBuyList: ArrayList<Product>
    ): Int {
        for (i in newBuyList.indices) {
            if (oldBuyList[i] != newBuyList[i]) {
                return i
            }
        }
        return 0
    }
}